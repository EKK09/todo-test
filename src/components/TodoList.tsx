import React, { useEffect } from 'react';
import useTodoListStore from 'src/store/todoList';
import styled from 'styled-components';
import TodoListItem from './TodoListItem';

const LoadingWrapper = styled.div`
  font-weight: bolder;
  display: flex;
  justify-content: center;
  align-items: center;
  flex-wrap: nowrap;
`;

const TodoList = (): React.ReactElement => {
  const { todoList, isFetching, fetchTodoList } = useTodoListStore(
    (state) => ({
      todoList: state.todoList,
      isFetching: state.isFetching,
      fetchTodoList: state.fetchTodoList,
    }),
  );
  const todoListItems = todoList.map((todo) => <TodoListItem key={todo.id} todo={todo} />);

  useEffect(() => {
    fetchTodoList();
  }, []);

  if (isFetching) {
    return <LoadingWrapper>Loading ....</LoadingWrapper>;
  }

  return (
    <div>
      {todoListItems}
    </div>
  );
};

export default TodoList;
