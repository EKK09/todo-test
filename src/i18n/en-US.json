{
    "global": {
        "appName": "TODO LIST",
        "logout": "Logout"
    },
    "loginPage": {
        "account": "Account",
        "password": "Password",
        "login": "Login"
    },
    "todo": {
        "completed": "Completed"
    }
}