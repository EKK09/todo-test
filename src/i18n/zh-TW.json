{
    "global": {
        "appName": "代辦事項"
    },
    "loginPage": {
        "account": "帳號",
        "password": "密碼",
        "login": "登入"
    },
    "todo": {
        "completed": "完成"
    }
}