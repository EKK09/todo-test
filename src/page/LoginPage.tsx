import React from 'react';
import { useHistory, useLocation } from 'react-router-dom';
import useUserStore from 'src/store/user';
import styled from 'styled-components';
import { Location } from 'history';

type LocationState = {
  from: Location;
};

const Wrapper = styled.div`
  max-width: 300px;
  margin: 60px auto;
`;

const InputWrapper = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  margin: 16px 0;
`;

const InputLabel = styled.span`
  font-weight: bolder;
  margin-right: 10px;
  width: 80px;
  display: inline-block;
`;
const Input = styled.input`
  flex: 1;
`;

const Button = styled.button`
  color: black;
  font-weight: bolder;
  background-color: white;
  border-radius: 4px;
  cursor: pointer;
  float: right;
`;

const LoginPage = (): React.ReactElement => {
  const setUserId = useUserStore((state) => state.setId);
  const history = useHistory();
  const location: Location<LocationState> = useLocation();

  const handleLogin = () => {
    // TODO: implement login api
    setUserId(1234);
    const { from } = location.state;
    const nextPath = from.pathname || '/';
    history.push(nextPath);
  };

  return (
    <Wrapper>
      <InputWrapper>
        <InputLabel>
          Account
        </InputLabel>
        <Input value="IamDavid" readOnly />
      </InputWrapper>
      <InputWrapper>
        <InputLabel>
          Password
        </InputLabel>
        <Input value="123456" type="password" readOnly />
      </InputWrapper>
      <Button onClick={handleLogin}>Login</Button>
    </Wrapper>
  );
};

export default LoginPage;
