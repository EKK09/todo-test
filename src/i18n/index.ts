import i18n from 'i18next';
import { initReactI18next } from 'react-i18next';
import enUSResource from './en-US.json';
import zhTWResource from './zh-TW.json';

export enum Lang {
    EN_US = 'en-US',
    ZH_TW = 'zh-TW',
}

export const LANG_OPTION = {
  [Lang.EN_US]: 'English',
  [Lang.ZH_TW]: '中文',
};

const resources = {
  [Lang.EN_US]: {
    translation: enUSResource,
  },
  [Lang.ZH_TW]: {
    translation: zhTWResource,
  },
};

const getDefaultLang = (locale: string): Lang => {
  switch (locale) {
    case Lang.EN_US:
      return Lang.EN_US;

    case Lang.ZH_TW:
      return Lang.ZH_TW;

    default:
      return Lang.EN_US;
  }
};

const locale = window.navigator.language;

i18n
  .use(initReactI18next)
  .init({
    resources,
    lng: getDefaultLang(locale),
  });

export default i18n;
