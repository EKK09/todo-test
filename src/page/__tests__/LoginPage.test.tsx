import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import TodoList from 'src/components/TodoList';
import { I18nextProvider } from 'react-i18next';
import i18n, { Lang } from 'src/i18n';
import { useHistory, useLocation } from 'react-router-dom';
import useUserStore from 'src/store/user';
import LoginPage from '../LoginPage';

jest.mock('react-router-dom', () => ({
  useHistory: jest.fn(),
  useLocation: jest.fn(),
}));

describe('LoginPage component', () => {
  describe('Render LoginPage', () => {
    it('chinese', async () => {
      i18n.changeLanguage(Lang.ZH_TW);
      const wrapper = render(
        <I18nextProvider i18n={i18n}>
          <LoginPage />
        </I18nextProvider>,
      );

      expect(wrapper.container).toMatchSnapshot();
    });
    it('english', async () => {
      i18n.changeLanguage(Lang.EN_US);
      const wrapper = render(
        <I18nextProvider i18n={i18n}>
          <LoginPage />
        </I18nextProvider>,
      );

      expect(wrapper.container).toMatchSnapshot();
    });
  });

  it('click login should set user id and push to from path', async () => {
    const mockSetId = jest.fn();
    const mockPush = jest.fn();
    const fooPath = '/fooPath';
    useUserStore.setState({ setId: mockSetId });
    (useHistory as jest.Mock).mockImplementation(() => ({ push: mockPush }));
    (useLocation as jest.Mock).mockImplementation(
      () => ({ state: { from: { pathname: fooPath } } }),
    );
    const wrapper = render(<LoginPage />);
    fireEvent.click(wrapper.getByText('Login'));

    // user id  is hard-coded here
    expect(mockSetId).toBeCalledWith(1234);
    expect(mockPush).toBeCalledWith(fooPath);
  });
});
