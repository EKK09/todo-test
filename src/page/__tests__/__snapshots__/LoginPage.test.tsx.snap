// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage component Render LoginPage chinese 1`] = `
<div>
  <div
    class="sc-bdfBQB dTpMfB"
  >
    <div
      class="sc-gsTEea dcpQyu"
    >
      <span
        class="sc-dlfnuX kVaApt"
      >
        帳號
      </span>
      <input
        class="sc-hKgJUU iQcqVk"
        readonly=""
        value="IamDavid"
      />
    </div>
    <div
      class="sc-gsTEea dcpQyu"
    >
      <span
        class="sc-dlfnuX kVaApt"
      >
        密碼
      </span>
      <input
        class="sc-hKgJUU iQcqVk"
        readonly=""
        type="password"
        value="123456"
      />
    </div>
    <button
      class="sc-eCstlR jfrbkj"
    >
      登入
    </button>
  </div>
</div>
`;

exports[`LoginPage component Render LoginPage english 1`] = `
<div>
  <div
    class="sc-bdfBQB dTpMfB"
  >
    <div
      class="sc-gsTEea dcpQyu"
    >
      <span
        class="sc-dlfnuX kVaApt"
      >
        Account
      </span>
      <input
        class="sc-hKgJUU iQcqVk"
        readonly=""
        value="IamDavid"
      />
    </div>
    <div
      class="sc-gsTEea dcpQyu"
    >
      <span
        class="sc-dlfnuX kVaApt"
      >
        Password
      </span>
      <input
        class="sc-hKgJUU iQcqVk"
        readonly=""
        type="password"
        value="123456"
      />
    </div>
    <button
      class="sc-eCstlR jfrbkj"
    >
      Login
    </button>
  </div>
</div>
`;
